x-pack/auditbeat/module/system/packages/rpm_linux.go

```diff
@@ -0,0 +1,303 @@
+//  int (*rpmReadConfigFiles)(const char*, const char*);
+//  rpmReadConfigFiles = (int (*)(const char*, const char*))f;
+//  return rpmReadConfigFiles(NULL, NULL);
+//}
+//
+//rpmdbMatchIterator
+//my_rpmtsInitIterator(void *f, rpmts ts) {
+//  rpmdbMatchIterator (*rpmtsInitIterator)(const rpmts, rpmTag, const void*, size_t);
+//  rpmtsInitIterator = (rpmdbMatchIterator (*)(const rpmts, rpmTag, const void*, size_t))f;
+//
+//  return rpmtsInitIterator(ts, RPMDBI_PACKAGES, NULL, 0);
+//}
+//
+//Header
+//my_rpmdbNextIterator(void *f, rpmdbMatchIterator mi) {
+//  Header (*rpmdbNextIterator)(rpmdbMatchIterator);
+//  rpmdbNextIterator = (Header (*)(rpmdbMatchIterator))f;
+//
+//  return rpmdbNextIterator(mi);
+//}
+//
+//Header
+//my_headerLink(void *f, Header h) {
+//  Header (*headerLink)(Header);
+//  headerLink = (Header (*)(Header))f;
+//
+//  return headerLink(h);
+//}
+//
+//int
+//my_headerGetEntry(void *f, Header h, rpm_tag_t tag, char **p) {
+//  int (*headerGetEntry)(Header, rpm_tag_t, rpm_tagtype_t*, rpm_data_t*, rpm_count_t*);
+//  headerGetEntry = (int (*)(Header, rpm_tag_t, rpm_tagtype_t*, rpm_data_t*, rpm_count_t*))f;
+//
+//  return headerGetEntry(h, tag, NULL, (void**)p, NULL);
+//}
+//
+//int
+//my_headerGetEntryInt(void *f, Header h, rpm_tag_t tag, int **p) {
+//  int (*headerGetEntry)(Header, rpm_tag_t, rpm_tagtype_t*, rpm_data_t*, rpm_count_t*);
+//  headerGetEntry = (int (*)(Header, rpm_tag_t, rpm_tagtype_t*, rpm_data_t*, rpm_count_t*))f;
+//
+//  return headerGetEntry(h, tag, NULL, (void**)p, NULL);
+//}
+//
+//void
+//my_headerFree(void *f, Header h) {
+//	Header (*headerFree)(Header);
+//  headerFree = (Header (*)(Header))f;
+//
+//	headerFree(h);
+//}
+//
+//void
+//my_rpmdbFreeIterator(void *f, rpmdbMatchIterator mi) {
+//  rpmdbMatchIterator (*rpmdbFreeIterator)(rpmdbMatchIterator);
+//  rpmdbFreeIterator = (rpmdbMatchIterator (*)(rpmdbMatchIterator))f;
+//
+//  rpmdbFreeIterator(mi);
+//}
+//
+//void
+//my_rpmtsFree(void *f, rpmts ts) {
+//  rpmts (*rpmtsFree)(rpmts);
+//  rpmtsFree = (rpmts (*)(rpmts))f;
+//
+//  rpmtsFree(ts);
+//}
+import "C"
+
+// Constants in sync with /usr/include/rpm/rpmtag.h
+const (
+	RPMTAG_NAME        = 1000
+	RPMTAG_VERSION     = 1001
+	RPMTAG_RELEASE     = 1002
+	RPMTAG_SUMMARY     = 1004
+	RPMTAG_LICENSE     = 1014
+	RPMTAG_URL         = 1020
+	RPMTAG_ARCH        = 1022
+	RPMTAG_SIZE        = 1009
+	RPMTAG_INSTALLTIME = 1008
+)
+
+type cFunctions struct {
+	rpmtsCreate        unsafe.Pointer
+	rpmReadConfigFiles unsafe.Pointer
+	rpmtsInitIterator  unsafe.Pointer
+	rpmdbNextIterator  unsafe.Pointer
+	headerLink         unsafe.Pointer
+	headerGetEntry     unsafe.Pointer
+	headerFree         unsafe.Pointer
+	rpmdbFreeIterator  unsafe.Pointer
+	rpmtsFree          unsafe.Pointer
+}
+
+var cFun *cFunctions
+
+func dlopenCFunctions() (*cFunctions, error) {
+	var librpmNames = []string{
+		"/usr/lib64/librpm.so",
+	}
+	var cFun cFunctions
+
+	librpm, err := dlopen.GetHandle(librpmNames)
```

Comment:
This handle is never [Closed()](https://godoc.org/github.com/coreos/pkg/dlopen#LibHandle.Close).

Reply:
Yes, it's not, but I cannot just add a `defer` because I cannot call the functions from it, it panics.  It's currently a singleton so it exists only once per process. Putting it in the Metricset structure would be possible, but requires a bigger refactoring of the whole module. Let me know if you think it's worth doing it anyway.